libtransmission/variant-converters.cc

```diff
@@ -0,0 +1,350 @@
+}
+
+///
+
+template<>
+std::optional<tr_port> VariantConverter::load<tr_port>(tr_variant* src)
+{
+    if (auto val = int64_t{}; tr_variantGetInt(src, &val))
+    {
+        return tr_port::fromHost(val);
+    }
+
+    return {};
+}
+
+template<>
+void VariantConverter::save<tr_port>(tr_variant* tgt, tr_port const& val)
+{
+    tr_variantInitInt(tgt, val.host());
+}
+
+///
+
+namespace PreallocationModeHelpers
+{
+// clang-format off
+static auto constexpr Keys = std::array<std::pair<std::string_view, tr_preallocation_mode>, 5>{{
+    { "off", TR_PREALLOCATE_NONE },
+    { "none", TR_PREALLOCATE_NONE },
+    { "fast", TR_PREALLOCATE_SPARSE },
+    { "sparse", TR_PREALLOCATE_SPARSE },
+    { "full", TR_PREALLOCATE_FULL },
+}};
+// clang-format on
+} // namespace PreallocationModeHelpers
+
+template<>
+std::optional<tr_preallocation_mode> VariantConverter::load<tr_preallocation_mode>(tr_variant* src)
+{
+    using namespace PreallocationModeHelpers;
+
+    if (auto val = std::string_view{}; tr_variantGetStrView(src, &val))
+    {
+        auto const needle = tr_strlower(tr_strvStrip(val));
+
+        for (auto const& [name, value] : Keys)
+        {
+            if (name == needle)
+            {
+                return value;
+            }
+        }
+    }
+
+    if (auto val = int64_t{}; tr_variantGetInt(src, &val))
+    {
+        for (auto const& [name, value] : Keys)
+        {
+            if (value == val)
+            {
+                return value;
+            }
+        }
+    }
+
+    return {};
+}
+
+template<>
+void VariantConverter::save<tr_preallocation_mode>(tr_variant* tgt, tr_preallocation_mode const& val)
+{
+    using namespace PreallocationModeHelpers;
+
+    for (auto const& [key, value] : Keys)
+    {
+        if (value == val)
+        {
+            tr_variantInitStrView(tgt, key);
+            return;
+        }
+    }
+}
+
+///
+
+template<>
+std::optional<size_t> VariantConverter::load<size_t>(tr_variant* src)
+{
+    if (auto val = int64_t{}; tr_variantGetInt(src, &val))
+    {
+        return static_cast<size_t>(val);
+    }
+
+    return {};
+}
+
+template<>
+void VariantConverter::save<size_t>(tr_variant* tgt, size_t const& val)
+{
+    tr_variantInitInt(tgt, val);
+}
+
+///
+
+template<>
+std::optional<std::string> VariantConverter::load<std::string>(tr_variant* src)
+{
+    if (auto val = std::string_view{}; tr_variantGetStrView(src, &val))
+    {
+        return std::string{ val };
+    }
+
+    return {};
+}
+
+template<>
+void VariantConverter::save<std::string>(tr_variant* tgt, std::string const& val)
+{
+    tr_variantInitStr(tgt, val);
+}
+
+///
+
+template<>
+std::optional<tr_tos_t> VariantConverter::load<tr_tos_t>(tr_variant* src)
+{
+    if (auto val = std::string_view{}; tr_variantGetStrView(src, &val))
+    {
+        return tr_tos_t::fromString(val);
+    }
+
+    if (auto val = int64_t{}; tr_variantGetInt(src, &val))
+    {
+        return tr_tos_t(val);
```

Comment:
```suggestion
        return tr_tos_t(static_cast<int>(val));
```

Reply:
Fixed in 2500751d83c91dc1cd4a6c1b00981d966c6c0b84